### PR TITLE
[inductor][pattern matcher] Retain meta tags

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102522
* #102450
* #102449
* #102578
* __->__ #102462
* #102448
* #102557
* #102547
* #102521

This will be used later on while propagating the `recompute` flag all the way to min-cut partitioner.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

### CODE REVIEW

File: torch/_inductor/pattern_matcher.py [hunk 1]
```diff
@@ -603,13 +606,13 @@ def run_node(self, node) -> Any:
                     result = graph.call_function(target, args, kwargs)
                     if "val" in node.meta and "val" not in result.meta:
                         result.meta["val"] = node.meta["val"]
+                    # Retain the meta tags from the first node in the match.
+                    # This is useful for retaining tags like recompute.
+                    result.meta.update(first_node.meta)
```
Comment (jansel) [inline torch/_inductor/pattern_matcher.py:611]: Do we need to skip updating meta["val"] here?

File: torch/_inductor/pattern_matcher.py [hunk 2]
```diff
@@ -603,13 +606,13 @@ def run_node(self, node) -> Any:
...
                     return result
                 raise NotImplementedError(f"unhandled {node}")
 
-        output_nodes = match.output_nodes()
-        node = output_nodes[0]
-
-        with graph.inserting_before(node):
+        with graph.inserting_before(first_node):
```
Comment (jansel) [inline torch/_inductor/pattern_matcher.py:615]: Is this change needed?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/pattern_matcher.py (hunk 2 above)
Comment (anijain2305) [reply torch/_inductor/pattern_matcher.py:623]: Oh, this is just variable name change.

**Review by jansel — APPROVED** [review]

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**